### PR TITLE
BigQuery: use autosummary to split up API reference docs

### PR DESCRIPTION
**Pull request by tswast**
@alixhami and I have heard from several folks that the [BigQuery API Reference docs](https://google-cloud-python.readthedocs.io/en/latest/bigquery/reference.html) are pretty impossible to navigate. This PR uses the [autosummary](http://www.sphinx-doc.org/en/master/ext/autosummary.html) module to split the docs into separate pages with an easier to browse summary table.

### CODE REVIEW

File: docs/bigquery/reference.rst [hunk 1]
```diff
@@ -1,68 +1,133 @@
...
+.. autosummary::
+    :toctree: generated
+
+    job.QueryJob
+    job.CopyJob
+    job.LoadJob
+    job.ExtractJob
+    job.UnknownJob
+
+Job Resources
```
Comment (tseaver) [inline docs/bigquery/reference.rst]: "Resources" doesn't seem right here:  it suggests things fetchable via an API request.  How about "Job-Related Types" or "Enumerations"?

File: third_party/sphinx/LICENSE [hunk 2]
```diff
@@ -0,0 +1,280 @@
+License for Sphinx
```
Comment (tseaver) [inline third_party/sphinx/LICENSE]: We ended up splitting out happybase because it was under a different license:  is there any consideration to that here?  I think that `sphinx-ext-autosummary` will use these templates by default from within its own resources:  are we customizing them here?

File: third_party/sphinx/sphinx/ext/autosummary/templates/autosummary/class.rst [hunk 3]
```diff
@@ -0,0 +1,33 @@
+{{ fullname | escape | underline}}
+
+.. currentmodule:: {{ module }}
+
+.. autoclass:: {{ objname }}
```
Comment (tseaver) [inline third_party/sphinx/sphinx/ext/autosummary/templates/autosummary/class.rst]: If we are customizing, then `:inherited-members:` should go here?

File: bigquery/google/cloud/bigquery/client.py [hunk 4]
```diff
@@ -603,7 +603,7 @@ def get_job(
         Returns:
             Union[google.cloud.bigquery.job.LoadJob, \
                   google.cloud.bigquery.job.CopyJob, \
-                  google.cloud.bigquery.job.ExtractJob \
+                  google.cloud.bigquery.job.ExtractJob, \
```
Comment (tseaver) [inline bigquery/google/cloud/bigquery/client.py]: These fixes are uncontroversial, as well as unrelated to the main thrust of this PR.  Maybe they should move to their own PR?

Comment (tswast) [reply bigquery/google/cloud/bigquery/client.py]: Good idea. Mailed https://github.com/GoogleCloudPlatform/google-cloud-python/pull/5342

File: third_party/sphinx/LICENSE (hunk 2 above)
Comment (tswast) [reply third_party/sphinx/LICENSE]: I believe it is okay in this case since we are putting it in a directory called `third_party`, which is different from the happybase case.

From some internal policies docs (go/thirdparty/non-google3):

> All non-Google code goes into the third_party/ directory of your project
>
> By default this should be a folder called third_party/ in the root directory of your project repository, but the exact name may vary by project.

File: third_party/sphinx/sphinx/ext/autosummary/templates/autosummary/class.rst (hunk 3 above)
Comment (tswast) [reply third_party/sphinx/sphinx/ext/autosummary/templates/autosummary/class.rst]: Yes, we are customizing slightly (the toctree so that sub-pages are generated). I'll add a comment.

Somehow that doesn't seem necessary for autosummary. I see the inherited members at https://storage.googleapis.com/tswast-docs-staging/google-cloud-python/pr-5340/bigquery/generated/google.cloud.bigquery.job.LoadJob.html#google.cloud.bigquery.job.LoadJob

File: docs/bigquery/reference.rst (hunk 1 above)
Comment (tswast) [reply docs/bigquery/reference.rst]: I'll use "Job-Related Types" to keep it open to non-enumerations if we ever add some.
